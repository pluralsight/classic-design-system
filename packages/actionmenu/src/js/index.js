import * as vars from '../vars/index.js'

export function calcNestedMenuPosition(parent, menu, origin) {
  return {
    topLeft: {
      left: -parent.width,
      top: `calc(-1 * ${vars.style.menuPaddingVert})`
    },
    topRight: {
      right: parent.width,
      top: `calc(-1 * ${vars.style.menuPaddingVert})`
    },
    bottomRight: {
      right: parent.width,
      top: -Math.abs(parent.height - menu.height)
    },
    bottomLeft: {
      left: -parent.width,
      top: -Math.abs(parent.height - menu.height)
    }
  }[origin]
}
