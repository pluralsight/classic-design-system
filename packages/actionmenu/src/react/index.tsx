import {
  RefForwardingComponent,
  ValueOf,
  classNames,
  dashify,
  handleMenuKeyDownEvents,
  handleMenuKeyUpEvents,
  useMenuRef,
  useResize,
  useClickOutside,
  useScrollOutside
} from '@pluralsight/ps-design-system-util'
import React from 'react'

import '../css/index.css'
import { ActionMenuContext } from './context'
import { Divider } from './divider'
import { Ellipsis } from './ellipsis'
import { Item } from './item'
import { Icon } from './icon'
import * as vars from '../vars/index'

// eslint-disable-next-line @typescript-eslint/no-empty-function
const noop = () => {}

interface ActionMenuStatics {
  Divider: typeof Divider
  Ellipsis: typeof Ellipsis
  Icon: typeof Icon
  Item: typeof Item
  origins: typeof vars.origins
  tagName: typeof vars.tagName
  useMenuRef: typeof useMenuRef
}
interface ActionMenuProps
  extends Omit<React.HTMLAttributes<HTMLUListElement>, 'onClick'> {
  onClick?: (evt: React.MouseEvent, value?: React.ReactText) => void
  onClose?: () => void
  origin?: ValueOf<typeof vars.origins>
}
type ActionMenuComponent = RefForwardingComponent<
  ActionMenuProps,
  HTMLUListElement,
  ActionMenuStatics
>
export const ActionMenu = React.forwardRef<HTMLUListElement, ActionMenuProps>(
  (
    {
      className,
      onClose = noop,
      origin = vars.origins.topLeft,
      onClick,
      children,
      ...rest
    },
    forwardedRef
  ) => {
    const ref = React.useRef<HTMLUListElement | null>(null)
    React.useImperativeHandle<HTMLUListElement | null, HTMLUListElement | null>(
      forwardedRef,
      () => ref.current
    )

    useClickOutside<HTMLUListElement>(ref, onClose)
    useResize<HTMLUListElement>(ref, onClose)
    useScrollOutside<HTMLUListElement>(ref, onClose)

    const handleKeyDown: React.KeyboardEventHandler<HTMLUListElement> = evt => {
      handleMenuKeyDownEvents(evt)
      if (evt.key === 'Escape') onClose()
    }

    return (
      <ul
        ref={ref}
        onKeyDown={handleKeyDown}
        onKeyUp={handleMenuKeyUpEvents}
        role="menu"
        {...rest}
        className={classNames(
          'psds-actionmenu',
          `psds-actionmenu--origin-${dashify(origin)}`,
          'psds-actionmenu__animation',
          className
        )}
      >
        <ActionMenuContext.Provider
          value={{ onClickContext: onClick, onClose, originContext: origin }}
        >
          {children}
        </ActionMenuContext.Provider>
      </ul>
    )
  }
) as ActionMenuComponent

ActionMenu.displayName = 'ActionMenu.Menu'

ActionMenu.Ellipsis = Ellipsis
ActionMenu.Icon = Icon
ActionMenu.Item = Item
ActionMenu.Divider = Divider
ActionMenu.origins = vars.origins
ActionMenu.tagName = vars.tagName

ActionMenu.useMenuRef = useMenuRef
export const origins = vars.origins

export default ActionMenu
