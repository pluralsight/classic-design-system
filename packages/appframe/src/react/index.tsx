import Button from '@pluralsight/ps-design-system-button'
import { breakpoints } from '@pluralsight/ps-design-system-core'
import Scrollable from '@pluralsight/ps-design-system-scrollable'
import Theme, { useTheme } from '@pluralsight/ps-design-system-theme'
import {
  RefForwardingComponent,
  ValueOf,
  canUseDOM,
  classNames,
  isFunction,
  useMatchMedia,
  usePrevious
} from '@pluralsight/ps-design-system-util'
import polyfillFocusWithin from 'focus-within'
import React from 'react'

import '../css/index.css'
import polyfillElementClosest from '../js/polyfill-element-closest'
import useBodyScrollLock from './use-body-scroll-lock'
import useOnClickOutside from './use-on-click-outside'
import useOnEscape from './use-on-escape'
import useOnInnerFocus from './use-on-inner-focus'
import * as vars from '../vars/index'

if (canUseDOM()) {
  polyfillElementClosest()
  polyfillFocusWithin(document)
}

const SKIP_TARGET_ID = 'ps-appframe--skip-target'
const TOP_NAV_ID = 'ps-appframe--topnav'

interface AppFrameContextValue {
  closeSidenav: () => void
  openSidenav: () => void
  sidenavOpen: boolean | undefined
  sidenavVariant: ValueOf<typeof vars.sidenavVariants>
}
const AppFrameContext = React.createContext<AppFrameContextValue>({
  closeSidenav: () => {},
  openSidenav: () => {},
  sidenavOpen: undefined,
  sidenavVariant: vars.sidenavVariants.closed
})

type RenderProp<P extends Record<string, unknown>> = (
  props: P
) => React.ReactNode

interface AppFrameProps extends React.HTMLAttributes<HTMLDivElement> {
  onRequestSideNavClose?: () => void
  onRequestSideNavOpen?: () => void
  sidenav?: React.ReactNode | RenderProp<{ visible: boolean }>
  sidenavOpen?: boolean
  topnav?:
    | React.ReactNode
    | RenderProp<
        Pick<
          AppFrameContextValue,
          'closeSidenav' | 'openSidenav' | 'sidenavOpen'
        >
      >
}
interface AppFrameStatics {
  SideNav: typeof SideNav
  TopNav: typeof TopNav
}

interface AppFrameComponent
  extends RefForwardingComponent<
    AppFrameProps,
    HTMLDivElement,
    AppFrameStatics
  > {}

const AppFrame = React.forwardRef((props, forwardedRef) => {
  const {
    children,
    className,
    onRequestSideNavClose,
    onRequestSideNavOpen,
    sidenav,
    sidenavOpen: initialSidenavOpen,
    topnav,
    ...rest
  } = props

  const ref = React.useRef<HTMLDivElement>(null)
  React.useImperativeHandle(
    forwardedRef,
    () => ref.current as unknown as HTMLDivElement
  )

  const themeName = useTheme()

  const smallMedia = useMatchMedia(`(min-width: ${breakpoints.small})`)
  const largeMedia = useMatchMedia(`(min-width: ${breakpoints.large})`)
  const prevLargeMedia = usePrevious(largeMedia)

  const defaultSidenavOpen = React.useMemo(() => {
    const controlled = typeof initialSidenavOpen !== 'undefined'
    return controlled ? initialSidenavOpen : largeMedia
  }, [initialSidenavOpen, largeMedia])

  const [sidenavOpen, setSidenavOpen] = React.useState(defaultSidenavOpen)

  const closeSidenav = React.useCallback(() => {
    const controlled = typeof props.sidenavOpen !== 'undefined'

    if (controlled && isFunction(onRequestSideNavClose)) onRequestSideNavClose()
    else if (!controlled) setSidenavOpen(false)
  }, [props.sidenavOpen, onRequestSideNavClose])

  const openSidenav = React.useCallback(() => {
    const controlled = typeof props.sidenavOpen !== 'undefined'

    if (controlled && isFunction(onRequestSideNavOpen)) onRequestSideNavOpen()
    else if (!controlled) setSidenavOpen(true)
  }, [props.sidenavOpen, onRequestSideNavOpen])

  React.useEffect(() => {
    const enteringXLarge = prevLargeMedia !== largeMedia && largeMedia
    if (enteringXLarge) openSidenav()
  }, [prevLargeMedia, largeMedia, openSidenav])

  React.useEffect(() => {
    const controlled = typeof props.sidenavOpen !== 'undefined'
    const next = controlled ? props.sidenavOpen : defaultSidenavOpen

    setSidenavOpen(next)
  }, [props.sidenavOpen, defaultSidenavOpen])

  const sidenavVariant = React.useMemo(() => {
    const { sidenavVariants: variants } = vars

    if (!sidenav) return variants.closed

    if (sidenavOpen) {
      return largeMedia ? variants.open : variants.overlay
    } else {
      return smallMedia ? variants.minimized : variants.closed
    }
  }, [sidenav, sidenavOpen, smallMedia, largeMedia])

  const skipTargetRef = React.useRef<HTMLAnchorElement>(null)

  const focusSkipTarget = React.useCallback(() => {
    if (!skipTargetRef.current) return

    skipTargetRef.current.focus()
  }, [])

  const contextValue = {
    closeSidenav,
    openSidenav,
    sidenavOpen,
    sidenavVariant
  }

  // TODO: className specs
  return (
    <AppFrameContext.Provider value={contextValue}>
      <div
        ref={ref}
        {...rest}
        className={classNames(
          'psds-appframe',
          `psds-theme--${themeName}`,
          className
        )}
      >
        <SkipBanner href={'#' + SKIP_TARGET_ID} />

        <TopNav id={TOP_NAV_ID}>{topnav}</TopNav>

        <Container>
          {sidenav && <SideNav>{sidenav}</SideNav>}

          <main className="psds-appframe__content">
            <SkipTarget
              id={SKIP_TARGET_ID}
              onClick={focusSkipTarget}
              ref={skipTargetRef}
            />

            {children}
          </main>
        </Container>
      </div>
    </AppFrameContext.Provider>
  )
}) as AppFrameComponent

AppFrame.displayName = 'AppFrame'

// TODO: why is there a an href here?
interface SkipBannerProps extends React.HTMLAttributes<HTMLDivElement> {
  href: string
}
const SkipBanner: React.FC<SkipBannerProps> = props => {
  return (
    <Theme name={Theme.names.dark}>
      <div {...props} className="psds-appframe__skip-banner">
        <Button
          appearance={Button.appearances.secondary}
          size={Button.sizes.small}
          tabIndex={0}
        >
          skip to main content
        </Button>
      </div>
    </Theme>
  )
}

interface SkipTargetProps extends React.HTMLAttributes<HTMLAnchorElement> {
  id: string
}
const SkipTarget = React.forwardRef<HTMLAnchorElement, SkipTargetProps>(
  (props, ref) => {
    return <a ref={ref} tabIndex={-1} {...props} />
  }
)

SkipTarget.displayName = 'SkipTarget'

const Container: React.FC<React.HTMLAttributes<HTMLDivElement>> = props => {
  const context = React.useContext(AppFrameContext)
  return (
    <div
      {...props}
      className={classNames(
        'psds-appframe__container',
        context.sidenavVariant &&
          `psds-appframe__container--${context.sidenavVariant}`
      )}
    />
  )
}

interface SideNavProps extends React.HTMLAttributes<HTMLDivElement> {
  children: AppFrameProps['sidenav']
}
interface SideNavStatics {
  variants: typeof vars.sidenavVariants
}
const SideNav: React.FC<SideNavProps> & SideNavStatics = ({
  className,
  ...rest
}) => {
  const { sidenavVariants: variants } = vars
  const children = rest.children as SideNavProps['children']

  const { closeSidenav, openSidenav, sidenavVariant } =
    React.useContext(AppFrameContext)

  const hoverable = sidenavVariant === variants.minimized
  const [hovered, setHovered] = React.useState(false)

  const variant = hoverable && hovered ? variants.overlay : sidenavVariant
  const visible = variant === variants.overlay || variant === variants.open

  const ref = React.useRef<HTMLDivElement>(null)

  const overlayed = variant === variants.overlay
  useBodyScrollLock(ref, overlayed)

  const closeIfOverlayed = React.useCallback(() => {
    if (overlayed) closeSidenav()
  }, [overlayed, closeSidenav])

  useOnClickOutside(ref, evt => {
    const target = evt.target as HTMLElement
    const inTopNav = target && !!target.closest(`#${TOP_NAV_ID}`)
    if (!inTopNav) closeIfOverlayed()
  })
  useOnEscape(closeIfOverlayed)
  useOnInnerFocus(ref, { onEnter: openSidenav, onLeave: closeIfOverlayed })

  return (
    <Theme name={Theme.names.dark}>
      <div
        ref={ref}
        {...rest}
        className={classNames(
          'psds-appframe__sidenav',
          variant && `psds-appframe__sidenav--${variant}`,
          className
        )}
        {...(hoverable && {
          onMouseEnter: () => setHovered(true),
          onMouseLeave: () => setHovered(false)
        })}
      >
        <div className="psds-appframe__sidenav__overflow-mask">
          <Scrollable className="psds-appframe__sidenav__inner">
            {typeof children === 'function' ? children({ visible }) : children}
          </Scrollable>
        </div>
      </div>
    </Theme>
  )
}

SideNav.variants = vars.sidenavVariants

interface TopNavProps extends React.HTMLAttributes<HTMLDivElement> {
  children: AppFrameProps['topnav']
}

const TopNav: React.FC<TopNavProps> = ({ className, ...rest }) => {
  const children = rest.children as TopNavProps['children']
  const { closeSidenav, openSidenav, sidenavOpen } =
    React.useContext(AppFrameContext)

  const meta = { closeSidenav, openSidenav, sidenavOpen }

  return (
    <Theme name={Theme.names.dark}>
      <div {...rest} className={classNames('psds-appframe__topnav', className)}>
        {isFunction<[typeof meta], React.ReactNode>(children)
          ? children(meta)
          : children}
      </div>
    </Theme>
  )
}

AppFrame.SideNav = SideNav
AppFrame.SideNav.displayName = 'AppFrame.SideNav'

AppFrame.TopNav = TopNav
AppFrame.TopNav.displayName = 'AppFrame.TopNav'

export default AppFrame
