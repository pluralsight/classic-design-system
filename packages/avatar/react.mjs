export { default } from './dist/esm/react/index'
export * from './dist/esm/vars/index'
export * from './dist/esm/js/index'
