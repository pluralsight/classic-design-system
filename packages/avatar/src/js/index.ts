import sha1 from 'tiny-hashes/sha1'

import { colors, fallbackPixel } from '../vars/index'

export const getInitials = (fullname?: string) => {
  if (!fullname || !fullname.length) return ':)'

  return fullname[0].toUpperCase()
}

export const getColorByName = (name?: string) => {
  const stableRandNumber = parseInt(sha1(name).replace(/[^0-9]/g, '')[0], 10)
  return colors[stableRandNumber] || colors[0]
}

export const transformSrc = (src = '') => {
  const hasGravatar = /gravatar/.exec(src)
  if (!hasGravatar) return src

  const sep = /\?/.exec(src) ? '&' : '?'
  const fallback = encodeURIComponent(fallbackPixel)

  return `${src}${sep}d=${fallback}`
}
