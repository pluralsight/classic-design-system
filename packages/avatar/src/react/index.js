import * as glamor from 'glamor'
import PropTypes from 'prop-types'
import React from 'react'

import css from '../css'
import { sizes } from '../vars'
import { getColorByName, getInitials, transformSrc } from './utils'

const styles = {
  avatar: ({ size }) =>
    glamor.css({
      ...css['.psds-avatar'],
      ...css[`.psds-avatar--size-${size}`]
    }),
  image: _ => glamor.css(css['.psds-avatar__image']),
  initials: ({ name }) => glamor.css(css['.psds-avatar__initials'])
}

class Avatar extends React.Component {
  constructor(props) {
    super(props)
    this.state = { imageState: 'loading' }
    this.handleImageLoadSuccess = this.handleImageLoadSuccess.bind(this)
    this.handleImageLoadError = this.handleImageLoadError.bind(this)
  }
  handleImageLoadSuccess() {
    this.setState({ imageState: 'success' })
  }
  handleImageLoadError() {
    this.setState({ imageState: 'error' })
  }
  render() {
    const { className, name, size, src, style } = this.props
    const { imageState } = this.state
    const avatarProps = {
      ...styles.avatar({ size }),
      ...(className ? { className } : null),
      ...(style ? { style } : null)
    }
    const shouldShowImg = src && imageState !== 'error'
    const shouldShowInitials =
      name && (imageState === 'error' || imageState === 'loading')

    return (
      <div {...avatarProps}>
        {shouldShowImg && (
          <img
            onError={this.handleImageLoadError}
            onLoad={this.handleImageLoadSuccess}
            {...styles.image()}
            src={transformSrc(src)}
          />
        )}
        {shouldShowInitials && (
          <div
            {...styles.initials({ name })}
            style={{ backgroundColor: getColorByName(name) }}
          >
            {getInitials(name)}
          </div>
        )}
      </div>
    )
  }
}

Avatar.defaultProps = {
  size: sizes.medium
}

Avatar.propTypes = {
  className: PropTypes.string,
  name: PropTypes.string,
  size: PropTypes.oneOf(Object.keys(sizes)),
  src: PropTypes.string
}

Avatar.sizes = sizes

export default Avatar
