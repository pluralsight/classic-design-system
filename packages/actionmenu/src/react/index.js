import { elementOfType } from '@pluralsight/ps-design-system-prop-types'

import PropTypes from 'prop-types'

import ActionMenu from './menu.js'
import Divider from './divider.js'
import Item from './item.js'

import * as vars from '../vars/index.js'

ActionMenu.Item = Item
ActionMenu.Divider = Divider
ActionMenu.origins = vars.origins
ActionMenu.propTypes = {
  children: PropTypes.oneOfType([
    elementOfType(Item),
    elementOfType(Divider),
    PropTypes.arrayOf(
      PropTypes.oneOfType([elementOfType(Item), elementOfType(Divider)])
    )
  ]),
  onClose: PropTypes.func
}
ActionMenu.defaultProps = {
  onClose: () => {}
}
export const origins = vars.origins

export default ActionMenu
