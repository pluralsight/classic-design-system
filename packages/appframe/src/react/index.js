import { canUseDOM } from 'exenv'
import { compose, css } from 'glamor'
import polyfillFocusWithin from 'focus-within'
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState
} from 'react'
import PropTypes from 'prop-types'

import filterReactProps from '@pluralsight/ps-design-system-filter-react-props'

import Button from '@pluralsight/ps-design-system-button'
import { breakpoints } from '@pluralsight/ps-design-system-core'
import Scrollable from '@pluralsight/ps-design-system-scrollable'
import Theme, { useTheme } from '@pluralsight/ps-design-system-theme'
import { isFunction, usePrevious } from '@pluralsight/ps-design-system-util'

import stylesheet from '../css/index.js'
import polyfillElementClosest from '../js/polyfill-element-closest.js'
import * as vars from '../vars/index.js'

import useBodyScrollLock from './use-body-scroll-lock.js'
import useMatchMedia from './use-match-media.js'
import useOnClickOutside from './use-on-click-outside.js'
import useOnEscape from './use-on-escape.js'
import useOnInnerFocus from './use-on-inner-focus.js'

if (canUseDOM) {
  polyfillElementClosest()
  polyfillFocusWithin(document)
}

const SKIP_TARGET_ID = 'ps-appframe--skip-target'
const TOP_NAV_ID = 'ps-appframe--topnav'

const styles = {
  appframe: (themeName, _props) =>
    compose(
      css(stylesheet['.psds-appframe']),
      css(stylesheet[`.psds-appframe.psds-theme--${themeName}`])
    ),

  skipBanner: () => css(stylesheet['.psds-appframe__skip-banner']),

  container: variant =>
    compose(
      css(stylesheet['.psds-appframe__container']),
      variant && css(stylesheet[`.psds-appframe__container--${variant}`])
    ),
  content: () => css(stylesheet['.psds-appframe__content']),
  sidenav: variant =>
    compose(
      css(stylesheet['.psds-appframe__sidenav']),
      variant && css(stylesheet[`.psds-appframe__sidenav--${variant}`])
    ),
  sidenavOverflowMask: () =>
    css(stylesheet['.psds-appframe__sidenav__overflow-mask']),
  sidenavInner: () => css(stylesheet['.psds-appframe__sidenav__inner']),
  topnav: () => css(stylesheet['.psds-appframe__topnav'])
}

const AppFrameContext = createContext()

const AppFrame = React.forwardRef((props, forwardedRef) => {
  const {
    children,
    onRequestSideNavClose,
    onRequestSideNavOpen,
    sidenav,
    topnav
  } = props

  const ref = React.useRef()
  useImperativeHandle(forwardedRef, () => ref.current)

  const themeName = useTheme()

  const smallMedia = useMatchMedia(`(min-width: ${breakpoints.small})`)
  const largeMedia = useMatchMedia(`(min-width: ${breakpoints.large})`)
  const prevLargeMedia = usePrevious(largeMedia)

  const defaultSidenavOpen = useMemo(() => {
    const controlled = typeof props.sidenavOpen !== 'undefined'
    return controlled ? props.sidenavOpen : largeMedia
  }, [props.sidenavOpen, largeMedia])

  const [sidenavOpen, setSidenavOpen] = useState(defaultSidenavOpen)

  const closeSidenav = useCallback(() => {
    const controlled = typeof props.sidenavOpen !== 'undefined'

    if (controlled && isFunction(onRequestSideNavClose)) onRequestSideNavClose()
    else if (!controlled) setSidenavOpen(false)
  }, [props.sidenavOpen, onRequestSideNavClose])

  const openSidenav = useCallback(() => {
    const controlled = typeof props.sidenavOpen !== 'undefined'

    if (controlled && isFunction(onRequestSideNavOpen)) onRequestSideNavOpen()
    else if (!controlled) setSidenavOpen(true)
  }, [props.sidenavOpen, onRequestSideNavOpen])

  useEffect(() => {
    const enteringXLarge = (prevLargeMedia !== largeMedia) & largeMedia
    if (enteringXLarge) openSidenav()
  }, [prevLargeMedia, largeMedia, openSidenav])

  useEffect(() => {
    const controlled = typeof props.sidenavOpen !== 'undefined'
    const next = controlled ? props.sidenavOpen : defaultSidenavOpen

    setSidenavOpen(next)
  }, [props.sidenavOpen, defaultSidenavOpen])

  const sidenavVariant = useMemo(() => {
    const { sidenavVariants: variants } = vars

    if (!sidenav) return variants.closed

    if (sidenavOpen) {
      return largeMedia ? variants.open : variants.overlay
    } else {
      return smallMedia ? variants.minimized : variants.closed
    }
  }, [sidenav, sidenavOpen, smallMedia, largeMedia])

  const skipTargetRef = useRef()

  const focusSkipTarget = useCallback(() => {
    if (!skipTargetRef.current) return

    skipTargetRef.current.focus()
  }, [])

  const contextValue = {
    closeSidenav,
    openSidenav,
    sidenavOpen,
    sidenavVariant
  }

  return (
    <AppFrameContext.Provider value={contextValue}>
      <div
        ref={ref}
        {...styles.appframe(themeName, props)}
        {...filterReactProps(props)}
      >
        <SkipBanner href={'#' + SKIP_TARGET_ID} />

        <TopNav id={TOP_NAV_ID}>{topnav}</TopNav>

        <Container>
          {sidenav && <SideNav>{sidenav}</SideNav>}

          <main {...styles.content()}>
            <SkipTarget
              id={SKIP_TARGET_ID}
              onClick={focusSkipTarget}
              ref={skipTargetRef}
            />

            {children}
          </main>
        </Container>
      </div>
    </AppFrameContext.Provider>
  )
})

AppFrame.displayName = 'AppFrame'
AppFrame.defaultProps = {}

AppFrame.propTypes = {
  children: PropTypes.node,
  onRequestSideNavClose: PropTypes.func,
  onRequestSideNavOpen: PropTypes.func,
  sidenav: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  sidenavOpen: PropTypes.bool,
  topnav: PropTypes.oneOfType([PropTypes.func, PropTypes.node])
}

function SkipBanner(props) {
  return (
    <Theme name={Theme.names.dark}>
      <div {...styles.skipBanner()} {...props}>
        <Button
          appearance={Button.appearances.secondary}
          size={Button.sizes.small}
          tabIndex={0}
        >
          skip to main content
        </Button>
      </div>
    </Theme>
  )
}

SkipBanner.propTypes = {
  href: PropTypes.string.isRequired
}

const SkipTarget = React.forwardRef((props, ref) => {
  return <a ref={ref} tabIndex={-1} {...props} />
})

SkipTarget.displayName = 'SkipTarget'
SkipTarget.propTypes = {
  id: PropTypes.string.isRequired
}

function Container(props) {
  const context = useContext(AppFrameContext)
  return <div {...styles.container(context.sidenavVariant)} {...props} />
}

function SideNav(props) {
  const { sidenavVariants: variants } = vars

  const { children, ...rest } = props
  const { closeSidenav, openSidenav, sidenavVariant } = useContext(
    AppFrameContext
  )

  const hoverable = sidenavVariant === variants.minimized
  const [hovered, setHovered] = useState(false)

  const variant = hoverable && hovered ? variants.overlay : sidenavVariant
  const visible = variant === variants.overlay || variant === variants.open

  const ref = React.useRef()

  const overlayed = variant === variants.overlay
  useBodyScrollLock(ref, overlayed)

  const closeIfOverlayed = useCallback(() => {
    if (overlayed) closeSidenav()
  }, [overlayed, closeSidenav])

  useOnClickOutside(ref, evt => {
    const inTopNav = evt.target && !!evt.target.closest(`#${TOP_NAV_ID}`)
    if (!inTopNav) closeIfOverlayed()
  })
  useOnEscape(closeIfOverlayed)
  useOnInnerFocus(ref, { onEnter: openSidenav, onLeave: closeIfOverlayed })

  return (
    <Theme name={Theme.names.dark}>
      <div
        ref={ref}
        {...styles.sidenav(variant)}
        {...rest}
        {...(hoverable && {
          onMouseEnter: () => setHovered(true),
          onMouseLeave: () => setHovered(false)
        })}
      >
        <div {...styles.sidenavOverflowMask()}>
          <Scrollable {...styles.sidenavInner()}>
            {isFunction(children) ? children({ visible }) : children}
          </Scrollable>
        </div>
      </div>
    </Theme>
  )
}

SideNav.propTypes = {
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.node])
}
SideNav.variants = vars.sidenavVariants

function TopNav(props) {
  const { children, ...rest } = props
  const { closeSidenav, openSidenav, sidenavOpen } = useContext(AppFrameContext)

  const meta = { closeSidenav, openSidenav, sidenavOpen }

  return (
    <Theme name={Theme.names.dark}>
      <div {...styles.topnav()} {...rest}>
        {isFunction(children) ? children(meta) : children}
      </div>
    </Theme>
  )
}
TopNav.propTypes = {
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.node])
}

AppFrame.SideNav = SideNav
AppFrame.SideNav.displayName = 'AppFrame.SideNav'

AppFrame.TopNav = TopNav
AppFrame.TopNav.displayName = 'AppFrame.TopNav'

export default AppFrame
