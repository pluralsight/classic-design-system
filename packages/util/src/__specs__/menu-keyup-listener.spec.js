import React, { createRef } from 'react'
import { fireEvent, render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { useMenuKeyEvents } from '../index.js'

const TestComponent = props => {
  const rootList = [...Array(10).keys()]
  const nestedList = [...Array(5).keys()]
  const ref = createRef(null)
  useMenuKeyEvents(ref)
  return (
    <ul data-testid="root-menu">
      {rootList.map((el, i) => {
        return i === 4 || i === 7 ? (
          <ul data-testid="sub-menu" key={`root-list-item-${el}`}>
            {nestedList.map(sub => (
              <li
                key={
                  i === 4
                    ? `first-sub-menu-list-item-${sub}`
                    : `second-sub-menu-list-item-${sub}`
                }
                data-testid={
                  i === 4
                    ? `first-sub-menu-list-item-${sub}`
                    : `second-sub-menu-list-item-${sub}`
                }
              >
                sub menu list item
              </li>
            ))}
          </ul>
        ) : (
          <li data-testid={`root-list-item-${el}`} key={`root-list-item-${el}`}>
            root list item
          </li>
        )
      })}
    </ul>
  )
}

describe('useMenuKeyEvents()', () => {
  it('TestComponent Snapshot', () => {
    const { getByTestId } = render(<TestComponent />)
    const root = getByTestId('root-menu')
    expect(root).toMatchSnapshot()
  })
  it('Key navigation of menu', () => {
    const { getByTestId } = render(<TestComponent />)
    // const root = getByTestId('root-menu')
    expect(document.activeElement).toBe('root-list-item-0')
  })
})
