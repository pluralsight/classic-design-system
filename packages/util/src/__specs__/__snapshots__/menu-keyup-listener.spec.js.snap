// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useMenuKeyEvents() TestComponent Snapshot 1`] = `
<ul
  data-testid="root-menu"
>
  <li
    data-testid="root-list-item-0"
  >
    root list item
  </li>
  <li
    data-testid="root-list-item-1"
  >
    root list item
  </li>
  <li
    data-testid="root-list-item-2"
  >
    root list item
  </li>
  <li
    data-testid="root-list-item-3"
  >
    root list item
  </li>
  <ul
    data-testid="sub-menu"
  >
    <li
      data-testid="first-sub-menu-list-item-0"
    >
      sub menu list item
    </li>
    <li
      data-testid="first-sub-menu-list-item-1"
    >
      sub menu list item
    </li>
    <li
      data-testid="first-sub-menu-list-item-2"
    >
      sub menu list item
    </li>
    <li
      data-testid="first-sub-menu-list-item-3"
    >
      sub menu list item
    </li>
    <li
      data-testid="first-sub-menu-list-item-4"
    >
      sub menu list item
    </li>
  </ul>
  <li
    data-testid="root-list-item-5"
  >
    root list item
  </li>
  <li
    data-testid="root-list-item-6"
  >
    root list item
  </li>
  <ul
    data-testid="sub-menu"
  >
    <li
      data-testid="second-sub-menu-list-item-0"
    >
      sub menu list item
    </li>
    <li
      data-testid="second-sub-menu-list-item-1"
    >
      sub menu list item
    </li>
    <li
      data-testid="second-sub-menu-list-item-2"
    >
      sub menu list item
    </li>
    <li
      data-testid="second-sub-menu-list-item-3"
    >
      sub menu list item
    </li>
    <li
      data-testid="second-sub-menu-list-item-4"
    >
      sub menu list item
    </li>
  </ul>
  <li
    data-testid="root-list-item-8"
  >
    root list item
  </li>
  <li
    data-testid="root-list-item-9"
  >
    root list item
  </li>
</ul>
`;
